File: Sources/MDifyCore/Services/ConversionService.swift
import Combine
import Foundation

@MainActor
public final class ConversionService: ObservableObject {
    @Published public private(set) var items: [ConversionItem] = []
    @Published public var selectedID: UUID?
    @Published public private(set) var isConverting = false

    private let workerClient: any WorkerConverting
    private let namer: OutputFileNamer
    private var shouldCancel = false

    public init(
        workerClient: any WorkerConverting = WorkerBundleResolver().makeNativeRoutingClient(),
        namer: OutputFileNamer = OutputFileNamer()
    ) {
        self.workerClient = workerClient
        self.namer = namer
    }

    public var selectedItem: ConversionItem? {
        guard let selectedID else { return items.first }
        return items.first(where: { $0.id == selectedID })
    }

    public func enqueue(files: [URL]) {
        let existing = Set(items.map(\.inputURL))
        let newItems = files
            .filter { $0.isFileURL && !existing.contains($0) }
            .map { ConversionItem(inputURL: $0) }
        items.append(contentsOf: newItems)
        if selectedID == nil {
            selectedID = items.first?.id
        }
    }

    @discardableResult
    public func enqueue(folderScan: FolderScanResult) -> FolderImportSummary {
        let existing = Set(items.map(\.inputURL))
        let scannedFiles = folderScan.files.filter { !existing.contains($0.url) }
        let newItems = scannedFiles.map {
            ConversionItem(
                inputURL: $0.url,
                sourceRootURL: folderScan.rootURL,
                relativeOutputPath: $0.relativePath
            )
        }
        items.append(contentsOf: newItems)
        if selectedID == nil {
            selectedID = items.first?.id
        }

        return FolderImportSummary(
            addedCount: newItems.count,
            skippedUnsupportedCount: folderScan.skippedUnsupportedCount,
            skippedDuplicateCount: folderScan.files.count - scannedFiles.count
        )
    }

    public func clearCompleted() {
        items.removeAll { $0.status == .succeeded || $0.status == .failed || $0.status == .cancelled }
        if let selectedID, !items.contains(where: { $0.id == selectedID }) {
            self.selectedID = items.first?.id
        }
    }

    public func removeItem(id: UUID) {
        items.removeAll { $0.id == id }
        if selectedID == id || selectedID == nil || !items.contains(where: { $0.id == selectedID }) {
            selectedID = items.first?.id
        }
    }

    public func cancel() {
        shouldCancel = true
    }

    public func convertAll(outputDirectory: URL) async {
        guard !isConverting else { return }
        isConverting = true
        shouldCancel = false
        defer { isConverting = false }

        var reservedNames = Set<String>()
        var reservedRoots: [String: URL] = [:]
        for item in items where item.status == .pending || item.status == .failed {
            if shouldCancel {
                update(item.id) { $0.status = .cancelled }
                continue
            }

            let outputURL: URL
            if item.sourceRootURL == nil {
                outputURL = namer.markdownURL(for: item.inputURL, in: outputDirectory, reserved: reservedNames)
                reservedNames.insert(outputURL.lastPathComponent)
            } else {
                outputURL = namer.markdownURL(for: item, in: outputDirectory, reservedRoots: &reservedRoots)
            }
            await convert(itemID: item.id, outputURL: outputURL)
        }
    }

    private func convert(itemID: UUID, outputURL: URL) async {
        guard let item = items.first(where: { $0.id == itemID }) else { return }
        update(itemID) {
            $0.status = .converting
            $0.errorMessage = nil
            $0.outputURL = outputURL
            $0.log = "Running embedded MDify worker for \(item.inputURL.path)"
        }

        do {
            let response = try await workerClient.convert(inputURL: item.inputURL, outputURL: outputURL)
            guard response.ok else {
                update(itemID) {
                    $0.status = .failed
                    $0.errorMessage = response.message ?? response.errorCode ?? "Conversion failed."
                    $0.log = response.errorCode ?? "Worker reported failure."
                }
                return
            }

            let markdown = try String(contentsOf: outputURL, encoding: .utf8)
            update(itemID) {
                $0.status = .succeeded
                $0.markdownText = markdown
                $0.outputURL = outputURL
                $0.log = workerLog(response)
            }
        } catch {
            update(itemID) {
                $0.status = .failed
                $0.errorMessage = error.localizedDescription
                $0.log = error.localizedDescription
            }
        }
    }

    private func workerLog(_ response: WorkerResponse) -> String {
        var lines = [
            "Worker: \(response.worker)",
            "Engine: \(response.engine ?? "unknown")",
            "OCR used: \(response.ocrUsed ? "yes" : "no")"
        ]
        if !response.warnings.isEmpty {
            lines.append("Warnings: \(response.warnings.joined(separator: "; "))")
        }
        return lines.joined(separator: "\n")
    }

    private func update(_ id: UUID, mutate: (inout ConversionItem) -> Void) {
        guard let index = items.firstIndex(where: { $0.id == id }) else { return }
        mutate(&items[index])
    }
}

File: windows/MDify.Windows/Services/ConversionService.cs
using System.Collections.ObjectModel;
using System.IO;
using MDify.Windows.Models;
using MDify.Windows.Support;

namespace MDify.Windows.Services;

public sealed class ConversionService
{
    private readonly IWorkerConverting _workerClient;
    private readonly OutputFileNamer _namer;

    public ConversionService(
        IWorkerConverting? workerClient = null,
        OutputFileNamer? namer = null,
        ConversionOptions? options = null)
    {
        _workerClient = workerClient ?? CreateDefaultWorkerClient();
        _namer = namer ?? new OutputFileNamer();
        Options = options ?? new ConversionOptions();
    }

    public ObservableCollection<ConversionItem> Items { get; } = new();

    public ConversionOptions Options { get; set; }

    public OcrLanguageMode OcrLanguage
    {
        get => Options.OcrLanguage;
        set => Options = new ConversionOptions(value);
    }

    public void EnqueueFiles(IEnumerable<string> filePaths)
    {
        var existing = Items.Select(item => item.InputPath).ToHashSet(StringComparer.OrdinalIgnoreCase);
        foreach (var filePath in filePaths.Where(File.Exists))
        {
            if (existing.Add(filePath))
            {
                Items.Add(new ConversionItem(filePath));
            }
        }
    }

    public int EnqueueFolderScan(FolderScanResult folderScan)
    {
        var existing = Items.Select(item => item.InputPath).ToHashSet(StringComparer.OrdinalIgnoreCase);
        var addedCount = 0;
        foreach (var file in folderScan.Files)
        {
            if (existing.Add(file.Path))
            {
                Items.Add(new ConversionItem(
                    file.Path,
                    sourceRootPath: folderScan.RootPath,
                    relativeOutputPath: file.RelativePath));
                addedCount++;
            }
        }

        return addedCount;
    }

    public void ClearCompleted()
    {
        for (var index = Items.Count - 1; index >= 0; index--)
        {
            if (Items[index].Status is ConversionStatus.Succeeded or ConversionStatus.Failed or ConversionStatus.Cancelled)
            {
                Items.RemoveAt(index);
            }
        }
    }

    public void RemoveItem(Guid id)
    {
        var index = IndexOf(id);
        if (index >= 0)
        {
            Items.RemoveAt(index);
        }
    }

    public async Task ConvertAllAsync(
        string outputDirectory,
        CancellationToken cancellationToken)
    {
        Directory.CreateDirectory(outputDirectory);
        var reservedPaths = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
        var batchOptions = Options;

        for (var index = 0; index < Items.Count; index++)
        {
            var item = Items[index];
            if (item.Status is not (ConversionStatus.Pending or ConversionStatus.Failed))
            {
                continue;
            }

            if (cancellationToken.IsCancellationRequested)
            {
                MarkRemainingCancelled(index);
                break;
            }

            var outputPath = _namer.ReserveMarkdownPath(item, outputDirectory, reservedPaths);
            await ConvertItemAsync(index, item, outputPath, batchOptions, cancellationToken);
        }
    }

    private async Task ConvertItemAsync(
        int index,
        ConversionItem item,
        string outputPath,
        ConversionOptions options,
        CancellationToken cancellationToken)
    {
        Items[index] = item with
        {
            Status = ConversionStatus.Converting,
            OutputPath = outputPath,
            ErrorMessage = null,
            Log = $"Running MDify worker for {item.InputPath}"
        };

        try
        {
            var response = await _workerClient.ConvertAsync(item.InputPath, outputPath, options, cancellationToken);
            if (!response.Ok)
            {
                Items[index] = Items[index] with
                {
                    Status = ConversionStatus.Failed,
                    ErrorMessage = response.Message ?? response.ErrorCode ?? "Conversion failed.",
                    Log = response.ErrorCode ?? "Worker reported failure."
                };
                return;
            }

            var markdown = await File.ReadAllTextAsync(outputPath, CancellationToken.None);
            Items[index] = Items[index] with
            {
                Status = ConversionStatus.Succeeded,
                MarkdownText = markdown,
                OutputPath = outputPath,
                Log = WorkerLog(response)
            };
        }
        catch (OperationCanceledException)
        {
            Items[index] = Items[index] with { Status = ConversionStatus.Cancelled };
            MarkRemainingCancelled(index + 1);
        }
        catch (Exception error)
        {
            Items[index] = Items[index] with
            {
                Status = ConversionStatus.Failed,
                ErrorMessage = error.Message,
                Log = error.Message
            };
        }
    }

    private void MarkRemainingCancelled(int startIndex)
    {
        for (var index = startIndex; index < Items.Count; index++)
        {
            if (Items[index].Status is ConversionStatus.Pending or ConversionStatus.Failed)
            {
                Items[index] = Items[index] with { Status = ConversionStatus.Cancelled };
            }
        }
    }

    private int IndexOf(Guid id)
    {
        for (var index = 0; index < Items.Count; index++)
        {
            if (Items[index].Id == id)
            {
                return index;
            }
        }

        return -1;
    }

    private static string WorkerLog(WorkerResponse response)
    {
        var lines = new List<string>
        {
            $"Worker: {response.Worker}",
            $"Engine: {response.Engine ?? "unknown"}",
            $"OCR used: {(response.OcrUsed ? "yes" : "no")}"
        };
        if (response.Warnings.Count > 0)
        {
            lines.Add($"Warnings: {string.Join("; ", response.Warnings)}");
        }

        return string.Join(Environment.NewLine, lines);
    }

    private static IWorkerConverting CreateDefaultWorkerClient()
    {
        var resolver = new WorkerBundleResolver();
#if MDIFY_WINDOWS_LITE
        var liteClient = resolver.CreateClient(WorkerKind.Lite);
        return new NativeOcrRoutingClient(
            WorkerKind.Lite,
            liteClient,
            nativeOcr: new WindowsNativeOcrService());
#else
        return resolver.CreateClient(WorkerKind.Ocr, ocrMode: WorkerOcrMode.Auto);
#endif
    }
}
